homeassistant/core.py

```diff
@@ -498,8 +498,11 @@ def async_add_hass_job(
                 hassjob.target = cast(Callable[..., _R], hassjob.target)
             task = self.loop.run_in_executor(None, hassjob.target, *args)
 
-        self._tasks.add(task)
-        task.add_done_callback(self._tasks.remove)
+        # Hold a reference to self._tasks since
+        # we will swap it out in async_stop
+        tasks = self._tasks
+        tasks.add(task)
+        task.add_done_callback(tasks.remove)
```

Comment:
This shouldn't be necessary as `tasks.remove` is a method bound to the `tasks` instance. When `self._tasks` is swapped out, it shouldn't point at the new remove function.

Reply:
When the remove callback fires we would have already removed the task from self._tasks when we swapped it out so it would raise

Reply:
More specifically

We do
`self._tasks = set()`

As soon as we return control to the event loop with the next `async with` / `await` 
`async with self.timeout.async_timeout(STAGE_1_SHUTDOWN_TIMEOUT):`

One of the previously running tasks could finish and try to remove itself from `self._tasks` and it will no longer be there unless its pointing to the original `self._tasks` which is now only reachable via the local var `running_tasks` or the closure variable  `tasks`